### PR TITLE
Compare errors when fuzzing

### PR DESCRIPTION
**Pull request by pelletier**
Ref: https://github.com/segmentio/encoding/issues/3

Running this briefly locally reported some errors. For example, given the input `["\b"]` (backspace char in string), `encoding/json` reports an error, but `segmentio/encoding/json` accepts it.

### CODE REVIEW

**Comment by pelletier** [comment]
Here is a broader sample of crashes: https://gist.github.com/pelletier/5b4079b63204651e26bd32438e01cce1

File: json/fuzz/fuzz.go [hunk 1]
```diff
@@ -33,12 +33,24 @@ func Fuzz(data []byte) int {
...
-			continue
-		}
-		if err := json.Unmarshal(data, v2); err != nil {
-			panic(err)
+		encodingErr := encodingJSON.Unmarshal(data, v1)
+		err := json.Unmarshal(data, v2)
+
+		if encodingErr != nil {
+			if err != nil {
+				// both implementations report an error
```
Comment (achille-roussel) [inline json/fuzz/fuzz.go]: Could we add a check like
```go
if reflect.TypeOf(encodingErr) != reflect.TypeOf(err) {
    panic(fmt.Errorf("error types mismatch: encoding/json=%s segmentio/encoding=%s", encodingErr, err))
}
```
?

**Comment by achille-roussel** [comment]
@pelletier I fixed all the issues that were detected by the fuzz test, the package seems to always be returning errors of the same type than `encoding/json` now.

Some error strings are still different because we use more detailed error messages in some cases and I didn't want to lose that.

**Comment by pelletier** [comment]
Great! Thanks for taking care of this. Changes seem to have some impact on the basic Unmarshal tests (this branch vs master):

```
benchcmp segmentio-encoding-json-master.txt segmentio-encoding-json.txt
benchmark                                  old ns/op     new ns/op     delta
BenchmarkMarshal/*json.codeResponse2       5296312       5271759       -0.46%
BenchmarkUnmarshal/*json.codeResponse2     8606423       9043968       +5.08%

benchmark                                  old MB/s     new MB/s     speedup
BenchmarkMarshal/*json.codeResponse2       366.38       368.09       1.00x
BenchmarkUnmarshal/*json.codeResponse2     225.47       214.56       0.95x

benchmark                                  old allocs     new allocs     delta
BenchmarkMarshal/*json.codeResponse2       0              0              +0.00%
BenchmarkUnmarshal/*json.codeResponse2     35             35             +0.00%

benchmark                                  old bytes     new bytes     delta
BenchmarkMarshal/*json.codeResponse2       0             0             +0.00%
BenchmarkUnmarshal/*json.codeResponse2     7705          7909          +2.65%
```

Still, down to merge it anyway, as we promise to error when encoding/json does.

**Comment by achille-roussel** [comment]
Yes I assumed it would, I think it’s due to the check that characters in a string are all printable ascii, I’m working on optimizing this. If I don’t have something working well by end of day let’s just go with the current version.

**Pull request merged**